File: android/electrode-reactnative-bridge/src/main/java/com/walmartlabs/electrode/reactnative/bridge/EventListenerProcessor.java
package com.walmartlabs.electrode.reactnative.bridge;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.walmartlabs.electrode.reactnative.bridge.helpers.Logger;
import com.walmartlabs.electrode.reactnative.bridge.util.BridgeArguments;

/**
 * Class that takes care of registering an event listener to the bridge.
 *
 * @param <T> Event payload type: Accepted types are Primitive wrappers and {@link Bridgeable}
 */

public class EventListenerProcessor<T> extends BridgeProcessor {
    private static final String TAG = EventListenerProcessor.class.getSimpleName();

    private final String eventName;
    private final Class<T> eventPayLoadClass;
    private final ElectrodeBridgeEventListener<T> eventListener;

    public EventListenerProcessor(@NonNull String eventName, @NonNull final Class<T> eventPayLoadClass, @NonNull final ElectrodeBridgeEventListener<T> eventListener) {
        this.eventName = eventName;
        this.eventPayLoadClass = eventPayLoadClass;
        this.eventListener = eventListener;
    }

    @Override
    public void execute() {
        ElectrodeBridgeEventListener<ElectrodeBridgeEvent> intermediateEventListener = new ElectrodeBridgeEventListener<ElectrodeBridgeEvent>() {
            @Override
            public void onEvent(@Nullable ElectrodeBridgeEvent bridgeEvent) {
                if (bridgeEvent == null) {
                    throw new IllegalArgumentException("bridgeEvent cannot be null, should never reach here");
                }

                Logger.d(TAG, "Processing final result for the event with payload bundle(%s)", bridgeEvent);

                T result = null;
                if (eventPayLoadClass != None.class) {
                    result = (T) BridgeArguments.generateObject(bridgeEvent.getData(), eventPayLoadClass);
                    result = (T) preProcessObject(result, eventPayLoadClass);
                }

                eventListener.onEvent(result);
            }
        };
        ElectrodeBridgeHolder.addEventListener(eventName, intermediateEventListener);
    }
}

File: ios/ElectrodeReactNativeBridge/ElectrodeReactNativeBridge/EventListenerProcessor.swift
//
//  EventListenerProcessor.swift
//  ElectrodeReactNativeBridge
//
//  Created by Cody Garvin on 3/2/17.
//  Copyright © 2017 Bharath Marulasiddappa. All rights reserved.
//

import UIKit

class EventListenerProcessor<T: Bridgeable>: NSObject {
    private let tag: String
    private let eventName: String
    private let eventPayloadClass: T.Type
    private let eventListener: ElectrodeEventListener
    
    init(eventName: String, eventPayloadClass: T.Type, eventListener: ElectrodeEventListener) {
        tag = String(describing: type(of: self))
        self.eventName = eventName
        self.eventPayloadClass = eventPayloadClass
        self.eventListener = eventListener
    }
    
    func execute() {
        
        // Build an intermediate eventlistener
        let listener = ElectrodeEventListenerImplementor(eventPayloadClass: self.eventPayloadClass) {
            [weak self] result in
            self?.eventListener.onEvent(result)
        }
        
        // Add the event listener to the bridge
        ElectrodeBridgeHolder.sharedInstance().registerEventListener(eventName, eventListener: listener)
    }
}

class ElectrodeEventListenerImplementor<T: Bridgeable>: NSObject, ElectrodeEventListener {
    
    private var executor: ((_ listener: Bridgeable?) -> ())? = nil
    private let eventPayloadClass: T.Type
    
    init(eventPayloadClass: T.Type, executor: ((_ listener: Bridgeable?) -> ())?) {
        self.eventPayloadClass = eventPayloadClass
        self.executor = executor
    }

    func onEvent(_ eventPayload: Any?) {
        
        print("Processing final result for the event with payload bundle (\(eventPayload))")
        let result = T.generateObject(eventPayload) as? Bridgeable
        
        if let executor = executor {
            executor(result)
        }
    }
}

/*
public class EventListenerProcessor<T> {
    private static final String TAG = EventListenerProcessor.class.getSimpleName();
    
    private final String eventName;
    private final Class<T> eventPayLoadClass;
    private final ElectrodeBridgeEventListener<T> eventListener;
    
    public EventListenerProcessor(@NonNull String eventName, @NonNull final Class<T> eventPayLoadClass, @NonNull final ElectrodeBridgeEventListener<T> eventListener) {
        this.eventName = eventName;
        this.eventPayLoadClass = eventPayLoadClass;
        this.eventListener = eventListener;
    }
    
    public void execute() {
        ElectrodeBridgeEventListener<Bundle> intermediateEventListener = new ElectrodeBridgeEventListener<Bundle>() {
            @Override
            public void onEvent(@Nullable Bundle eventPayload) {
                Logger.d(TAG, "Processing final result for the event with payload bundle(%s)", eventPayload);
                T result = BridgeArguments.generateObject(eventPayload, eventPayLoadClass, BridgeArguments.Type.EVENT);
                eventListener.onEvent(result);
            }
        };
        ElectrodeBridgeHolder.addEventListener(eventName, intermediateEventListener);
    }
}
*/
